File: leetcode/p00038/p00038.go
package p00038

import (
	"fmt"
	"strings"
)

// Time Complexity: O(n^2)
// Space Complexity: O(n)
func countAndSay_byte(n int) string {
	if n == 1 {
		return "1"
	}

	tmp := countAndSay_byte(n - 1)

	var res strings.Builder
	keep := tmp[0]
	count := 1
	for i := 1; i < len(tmp); i++ {
		if tmp[i] != keep {
			res.WriteByte(byte(count + 48))
			res.WriteByte(keep)
			keep = tmp[i]
			count = 1
		} else {
			count++
		}
	}
	res.WriteByte(byte(count + 48))
	res.WriteByte(keep)

	return res.String()
}

// Time Complexity: O(n^2)
// Space Complexity: O(n)
func countAndSay_bruteforce(n int) string {
	if n == 1 {
		return "1"
	}

	tmp := countAndSay_bruteforce(n - 1)

	res := ""
	keep := tmp[0]
	count := 1
	for i := 1; i < len(tmp); i++ {
		if tmp[i] != keep {
			res = fmt.Sprintf("%s%d%c", res, count, keep)
			keep = tmp[i]
			count = 1
		} else {
			count++
		}
	}
	res = fmt.Sprintf("%s%d%c", res, count, keep)

	return res
}

File: leetcode/p00038/p00038.py
class Solution:
    # Time Complexity: O(n^2)
    # Space Complexity: O(n)
    def countAndSay(self, n: int) -> str:
        s = "1"
        if n == 1:
            return s
        
        for i in range(1, n):
            tmp = ""
            
            keep = s[0]
            count = 1
            for ch in range(1, len(s)):
                if s[ch] == keep:
                    count += 1
                else:
                    tmp += str(count) + keep
                    keep = s[ch]
                    count = 1
            tmp += str(count) + keep
            s = tmp
        
        return s

if __name__ == '__main__':
    s = Solution()

    assert s.countAndSay(1) == "1"
    assert s.countAndSay(4) == "1211"